qubespdfconverter/client.py

```diff
@@ -0,0 +1,675 @@
+        :param depth: Conversion queue size
+        :param in_place: Value of --in-place flag
+        """
+        self.batch = asyncio.Queue(depth)
+
+        publish_task = asyncio.create_task(self._publish(proc, bar))
+        consume_task = asyncio.create_task(self._consume())
+
+        try:
+            await asyncio.gather(publish_task, consume_task)
+        finally:
+            if not publish_task.done():
+                await cancel_task(publish_task)
+
+            if not consume_task.done():
+                await cancel_task(consume_task)
+
+            while not self.batch.empty():
+                batch_e = await self.batch.get()
+                await cancel_task(batch_e.task)
+                self.batch.task_done()
+
+
+    async def _publish(self, proc, bar):
+        """Receive initial representations and start their conversions"""
+        pages = []
+
+        for page in range(1, self.pagenums + 1):
+            rep = Representation(Path(self.pdf.parent, str(page)), "rgb", "png")
+            await rep.receive(proc)
+
+            task = asyncio.create_task(rep.convert(bar))
+            batch_e = BatchEntry(task, rep)
+
+            try:
+                await self.batch.put(batch_e)
+            except asyncio.CancelledError:
+                await cancel_task(task)
+                raise
+
+            pages.append(page)
+
+            if page % self.batch.maxsize == 0 or page == self.pagenums:
+                await self.batch.join()
+                await self._save_reps(pages)
+                pages = []
+
+
+    async def _consume(self):
+        """Convert initial representations to final form and save as PDF"""
+        for _ in range(1, self.pagenums + 1):
+            batch_e = await self.batch.get()
+            await batch_e.task
+            self.batch.task_done()
+
+
+    async def _save_reps(self, pages):
+        """Save final representations to a PDF file"""
+        images = []
+
+        for page in pages:
+            try:
+                images.append(await asyncio.get_running_loop().run_in_executor(
+                    None,
+                    Image.open,
+                    Path(self.pdf.parent, f"{page}.png"))
+                )
+            except IOError as e:
+                for image in images:
+                    await asyncio.get_running_loop().run_in_executor(
+                        None,
+                        image.close
+                    )
+                raise RepresentationError("Failed to open representation") from e
+
+        try:
+            await asyncio.get_running_loop().run_in_executor(
+                None,
+                functools.partial(images[0].save,
+                                  self.pdf,
+                                  "PDF",
+                                  resolution=100,
+                                  append=self.pdf.exists(),
+                                  append_images=[] if len(images) == 1 else images[1:],
```

Comment:
I think `append_images=images[1:]` would work in a single-page case too.

Reply:
How silly of me to assume that that wouldn't work.